ddtrace/contrib/pymongo/patch.py

```diff
@@ -1,14 +1,64 @@
 import pymongo
 
-from .client import TracedMongoClient
+from ddtrace import Pin
+from ddtrace.vendor.wrapt import wrap_function_wrapper as _w
+from ...utils.wrappers import unwrap as _u
+
+from ...ext import mongo as mongox, SpanTypes
+from ...utils.deprecation import deprecated
+from ...constants import SPAN_MEASURED_KEY
+from .client import TracedMongoClient, set_address_tags
 
 # Original Client class
 _MongoClient = pymongo.MongoClient
 
 
 def patch():
+    patch_pymongo_module()
+    # We should progressively get rid of TracedMongoClient. We now try to
+    # wrap methods individually. cf issue #_
     setattr(pymongo, 'MongoClient', TracedMongoClient)
 
 
 def unpatch():
+    unpatch_pymongo_module()
     setattr(pymongo, 'MongoClient', _MongoClient)
+
+
+@deprecated(message='Use patching instead (see the docs).', version='1.0.0')
+def trace_mongo_client(client, tracer, service=mongox.SERVICE):
+    traced_client = TracedMongoClient(client)
+    Pin(service=service, tracer=tracer).onto(traced_client)
+    return traced_client
+
+
+def patch_pymongo_module():
+    if getattr(pymongo, '_datadog_patch', False):
+        return
+    setattr(pymongo, '_datadog_patch', True)
+    Pin(service=mongox.SERVICE, app=mongox.SERVICE).onto(pymongo.pool.Pool)
+
+    # This method is called when pymongo creates a new TCP connection
+    _w("pymongo.pool", "Pool.connect", traced_pool_connect)
+
+
+def unpatch_pymongo_module():
+    if not getattr(pymongo, '_datadog_patch', False):
+        return
+    setattr(pymongo, '_datadog_patch', False)
+
+    _u(pymongo.pool.Pool, "connect")
+
+
+def traced_pool_connect(wrapped, instance, args, kwargs):
+    pin = Pin._find(wrapped, instance)
+    is_monitoring_pool = instance.opts.connect_timeout == instance.opts.socket_timeout
```

Comment:
Hey! Sorry for the long delay. Hmmm interesting. After looking at the examples (thanks for the code references!) it does seem like patching `_create_pool_for_server` would make sense. IMO it'd be better if we're able to avoid patching the undesired path entirely. Is it relatively easy to accomplish?

Reply:
No worries :) Once again, thanks for your time @Kyle-Verhoog 

> IMO it'd be better if we're able to avoid patching the undesired path entirely. Is it relatively easy to accomplish?

Agree with you but tbh I have not found a way to do it :/ Basically in pseudo-python, the flow is like this:

```python
class PyMongoClient:
    def __init__(self, ip, port):
        self.ip = ip
        self.port = port
        # This runs in a separate thread
        self.start_monitoring()

    def start_monitoring(self):
        while True:
            for monitor in self.monitors:
                monitor.get_socket(self.ip, self.port) as sock_info:
            	    sock_info.health_check()
                time.sleep(x)

    def run_cmd(self, cmd):
        server = self.servers[(self.ip, self.port)]
        with server.get_socket(self.ip, self.port) as sock_info:
            sock_info.send(cmd)

class Server:
    def get_socket(self):
        return self.pool.get_socket()

class Monitor:
    def get_socket(self);
        return self.pool.get_socket()

class Pool:
    def get_socket(self):
        # (...)
        try:
            # Lookup of an already existing, ready socket. This call does not take time
            socket = self.sockets.popleft()
        except IndexError:
            # Creation of a new socket+handshake. This call does take time
            socket = self.connect()
        return socket
```

So basically:
  - If we patch `pymongo.server.Server`'s `get_socket`, we actually trace both `Pool`'s `self.sockets.popleft()` and  `Pool`'s `self.connect()`. The second one is effectively what we are trying to trace while the first one is not a network call but a lookup in memory which we do not want to trace (I think?)
  - If we patch `pymongo.pool.Pool`'s `self.connect()`, we only trace actual network calls but we trace network calls of both the Monitor object and the Server object. That means that we generate traces for both the pymongo thread that handle incoming requests (good) but also from the pymongo thread that performs monitoring health_checks on the remote address (bad)

I do not see a way to only patch the critical path, unfortunately :( That leaves us with 2 viable options imo:
  - Trace `pymongo.pool.Pool`'s `get_socket`: it will generate tiny traces when the socket already exists and classic traces when the socket needs to be created
  - Do the hack which adds an hidden attribute pool instances that belong to `Servers`: it will generate traces only for actual network calls linked to pymongo direct requests but well, it's a bit of a hack

Reply:
Thanks for that write-up! It's actually really helpful as I'd forgotten some of the context 🙂.

So I'm okay with patching `get_socket` in general even if there's the chance that it's a quick memory look up. It's meaningful sometimes which is enough reason IMO and our overhead shouldn't be too bad (unless this is done a crazy amount of times). I think we should avoid instrumenting the health checks from Monitor.

> Trace `pymongo.pool.Pool`'s `get_socket`: it will generate tiny traces when the socket already exists and classic traces when the socket needs to be created

Did you mean `Server.get_socket` here? Else we'd still be getting the health checks? If you did mean `Server.get_socket` I think that sounds like the way to go :+1:. What do you think?

Reply:
> Did you mean Server.get_socket here?

Yep my bad, messed things up! Alright I'm changing the code to patch `Server.get_socket` :). I'll update you when it's done!

Reply:
Haha no problem. And great! I'm excited to get this merged 😄

Reply:
Hmm I've ran into a problem implementing that. Actually `Pool`'s `get_socket`'s code creates&connects the socket and then yields it, making `Server`'s `get_socket`'s a generator. Thus, we cannot trace it effectively with `wrap_function_wrapper` since generator functions instantly return a generator and the actual code is executed when the parent function decides to consume it.

I believe a workaround could be to consume the socket from the generator in our patch and then yield the socket but I don't think it is a good idea. Are there some wrappers to work with generators in ddtrace?

Reply:
Agh, I see. Hmm we don't have any utils for patching generators unfortunately. I think we're okay to intercept the generator and yield the socket ourselves as you described. We just unfortunately can't make use of `yield from` to maintain compatibility with Python 2.

What are your thoughts?